File: desktop-app-vue/src/main/ai-engine/cowork/skills/skill-md-parser.js
/**
 * SkillMdParser - SKILL.md 解析器
 *
 * 解析 Clawdbot 风格的 SKILL.md 文件，支持 YAML frontmatter 和 Markdown 正文。
 *
 * @module ai-engine/cowork/skills/skill-md-parser
 */

const fs = require("fs");
const path = require("path");
const { logger } = require("../../../utils/logger.js");

// 尝试加载 gray-matter，如果不存在则使用简易解析
let matter;
try {
  matter = require("gray-matter");
} catch {
  matter = null;
}

/**
 * 简易 YAML frontmatter 解析器（作为 gray-matter 的后备）
 * @param {string} content - 文件内容
 * @returns {{data: object, content: string}}
 */
function simpleFrontmatterParser(content) {
  const lines = content.split("\n");

  if (lines[0].trim() !== "---") {
    return { data: {}, content };
  }

  let endIndex = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === "---") {
      endIndex = i;
      break;
    }
  }

  if (endIndex === -1) {
    return { data: {}, content };
  }

  const yamlLines = lines.slice(1, endIndex);
  const body = lines
    .slice(endIndex + 1)
    .join("\n")
    .trim();

  // 简易 YAML 解析（支持嵌套对象）
  const data = {};
  const stack = [{ obj: data, indent: -1 }];
  let currentKey = null;
  let currentArray = null;

  for (const line of yamlLines) {
    // 跳过空行和注释
    if (!line.trim() || line.trim().startsWith("#")) {
      continue;
    }

    // 计算缩进
    const indent = line.length - line.trimStart().length;
    const trimmed = line.trim();

    // 数组项（以 - 开头）
    if (trimmed.startsWith("- ")) {
      const value = trimmed
        .slice(2)
        .trim()
        .replace(/^['"]|['"]$/g, "");
      if (currentArray) {
        currentArray.push(value);
      }
      continue;
    }

    // 键值对
    const colonIndex = trimmed.indexOf(":");
    if (colonIndex > 0) {
      const key = trimmed.slice(0, colonIndex).trim();
      let value = trimmed.slice(colonIndex + 1).trim();

      // 调整栈，找到正确的父对象
      while (stack.length > 1 && stack[stack.length - 1].indent >= indent) {
        stack.pop();
      }
      const parent = stack[stack.length - 1].obj;

      // 嵌套对象开始（value 为空）
      if (value === "") {
        const newObj = {};
        parent[key] = newObj;
        stack.push({ obj: newObj, indent });
        currentKey = null;
        currentArray = null;
        continue;
      }

      // 数组开始（空数组标记）
      if (value === "[]") {
        currentKey = key;
        currentArray = [];
        parent[key] = currentArray;
        continue;
      }

      // 内联数组 [a, b, c]
      if (value.startsWith("[") && value.endsWith("]")) {
        const items = value
          .slice(1, -1)
          .split(",")
          .map((s) => s.trim().replace(/^['"]|['"]$/g, ""))
          .filter((s) => s);
        parent[key] = items;
        currentKey = null;
        currentArray = null;
        continue;
      }

      // 布尔值
      if (value === "true") {
        value = true;
      } else if (value === "false") {
        value = false;
      }
      // 数字
      else if (/^-?\d+(\.\d+)?$/.test(value)) {
        value = parseFloat(value);
      }
      // 字符串（移除引号）
      else {
        value = value.replace(/^['"]|['"]$/g, "");
      }

      parent[key] = value;
      currentKey = key;
      currentArray = null;
    }
  }

  return { data, content: body };
}

/**
 * SkillMdParser 类
 */
class SkillMdParser {
  constructor(options = {}) {
    this.options = {
      // 最大描述长度
      maxDescriptionLength: options.maxDescriptionLength || 200,
      // 是否严格验证
      strictValidation: options.strictValidation !== false,
      ...options,
    };
  }

  /**
   * 解析 SKILL.md 文件
   * @param {string} filePath - 文件路径
   * @returns {Promise<object>} SkillDefinition
   */
  async parseFile(filePath) {
    try {
      const content = await fs.promises.readFile(filePath, "utf-8");
      return this.parseContent(content, filePath);
    } catch (error) {
      throw new Error(
        `Failed to parse SKILL.md at ${filePath}: ${error.message}`,
      );
    }
  }

  /**
   * 解析 SKILL.md 内容
   * @param {string} content - 文件内容
   * @param {string} sourcePath - 源文件路径
   * @returns {object} SkillDefinition
   */
  parseContent(content, sourcePath = "unknown") {
    // 使用 gray-matter 或简易解析器
    const parsed = matter ? matter(content) : simpleFrontmatterParser(content);

    const frontmatter = parsed.data || {};
    const body = parsed.content || "";

    // 规范化字段名（支持 kebab-case 和 camelCase）
    const normalized = this._normalizeFields(frontmatter);

    // 构建 SkillDefinition
    const definition = {
      // 必填字段
      name: normalized.name || path.basename(path.dirname(sourcePath)),
      description: normalized.description || "",

      // 来源信息
      source: normalized.source || "unknown",
      sourcePath,

      // 显示信息
      displayName: normalized.displayName || normalized.name || "",
      version: normalized.version || "1.0.0",
      category: normalized.category || "custom",
      tags: normalized.tags || [],

      // 调用控制
      userInvocable: normalized.userInvocable !== false,
      hidden: normalized.hidden === true,

      // 门控条件
      requires: {
        bins: normalized.requires?.bins || [],
        env: normalized.requires?.env || [],
      },
      os: normalized.os || ["win32", "darwin", "linux"],
      enabled: normalized.enabled !== false,

      // 执行配置
      handler: normalized.handler || null,
      capabilities: normalized.capabilities || [],
      supportedFileTypes: normalized.supportedFileTypes || [],

      // Markdown 正文
      body: body.trim(),

      // 原始 frontmatter
      _raw: frontmatter,
    };

    // 验证
    const validation = this.validate(definition);
    if (!validation.valid) {
      if (this.options.strictValidation) {
        throw new Error(`Invalid SKILL.md: ${validation.errors.join(", ")}`);
      }
      logger.warn(
        `[SkillMdParser] Validation warnings for ${sourcePath}: ${validation.errors.join(", ")}`,
      );
    }

    return definition;
  }

  /**
   * 验证 SkillDefinition
   * @param {object} definition - 技能定义
   * @returns {{valid: boolean, errors: string[]}}
   */
  validate(definition) {
    const errors = [];

    // 必填字段检查
    if (!definition.name) {
      errors.push("name is required");
    } else if (!/^[a-z][a-z0-9-]*$/i.test(definition.name)) {
      errors.push("name must be alphanumeric with hyphens (e.g., my-skill)");
    }

    if (!definition.description) {
      errors.push("description is required");
    } else if (
      definition.description.length > this.options.maxDescriptionLength
    ) {
      errors.push(
        `description exceeds ${this.options.maxDescriptionLength} characters`,
      );
    }

    // 版本格式检查
    if (definition.version && !/^\d+\.\d+\.\d+/.test(definition.version)) {
      errors.push("version should follow semver format (e.g., 1.0.0)");
    }

    // handler 路径检查（如果指定）
    if (definition.handler && !definition.handler.startsWith("./")) {
      errors.push("handler path should be relative (start with ./)");
    }

    // 平台检查
    const validPlatforms = ["win32", "darwin", "linux"];
    if (definition.os) {
      for (const platform of definition.os) {
        if (!validPlatforms.includes(platform)) {
          errors.push(`invalid platform: ${platform}`);
        }
      }
    }

    return {
      valid: errors.length === 0,
      errors,
    };
  }

  /**
   * 规范化字段名
   * @private
   */
  _normalizeFields(obj) {
    const result = {};

    const fieldMap = {
      name: "name",
      description: "description",
      "display-name": "displayName",
      displayName: "displayName",
      version: "version",
      category: "category",
      tags: "tags",
      "user-invocable": "userInvocable",
      userInvocable: "userInvocable",
      hidden: "hidden",
      requires: "requires",
      os: "os",
      enabled: "enabled",
      handler: "handler",
      capabilities: "capabilities",
      "supported-file-types": "supportedFileTypes",
      supportedFileTypes: "supportedFileTypes",
    };

    for (const [key, value] of Object.entries(obj)) {
      const normalizedKey = fieldMap[key] || key;
      result[normalizedKey] = value;
    }

    return result;
  }
}

module.exports = { SkillMdParser };

File: android-app/feature-ai/src/main/java/com/chainlesschain/android/feature/ai/cowork/skills/loader/SkillMdParser.kt
package com.chainlesschain.android.feature.ai.cowork.skills.loader

import com.chainlesschain.android.feature.ai.cowork.skills.model.*
import org.yaml.snakeyaml.Yaml
import timber.log.Timber

/**
 * Parser for SKILL.md files with YAML frontmatter + Markdown body.
 *
 * Format:
 * ```
 * ---
 * name: code-review
 * description: Review code quality
 * category: development
 * ...
 * ---
 *
 * # Code Review Skill
 *
 * Markdown instructions body...
 * ```
 */
class SkillMdParser {

    private val yaml = Yaml()

    /**
     * Parse a SKILL.md file content into a Skill object.
     *
     * @param content Raw file content
     * @param source  Which layer this skill comes from
     * @param sourcePath Original file path (for debugging)
     * @return Parsed Skill, or null if parsing fails
     */
    fun parse(content: String, source: SkillSource, sourcePath: String = ""): Skill? {
        return try {
            val (frontmatter, body) = splitFrontmatter(content)
            if (frontmatter.isBlank()) {
                Timber.w("SkillMdParser: No YAML frontmatter found in $sourcePath")
                return null
            }

            val yamlData = parseYaml(frontmatter)
            val metadata = buildMetadata(yamlData)
            val examples = extractExamples(body)

            Skill(
                metadata = metadata,
                instructions = body.trim(),
                examples = examples,
                source = source,
                sourcePath = sourcePath,
                enabled = yamlData["enabled"] as? Boolean ?: true
            )
        } catch (e: Exception) {
            Timber.e(e, "SkillMdParser: Failed to parse $sourcePath")
            null
        }
    }

    /**
     * Split content into YAML frontmatter and Markdown body.
     */
    internal fun splitFrontmatter(content: String): Pair<String, String> {
        val lines = content.lines()
        if (lines.isEmpty() || lines[0].trim() != "---") {
            return "" to content
        }

        var endIndex = -1
        for (i in 1 until lines.size) {
            if (lines[i].trim() == "---") {
                endIndex = i
                break
            }
        }

        if (endIndex == -1) {
            return "" to content
        }

        val frontmatter = lines.subList(1, endIndex).joinToString("\n")
        val body = lines.subList(endIndex + 1, lines.size).joinToString("\n")
        return frontmatter to body
    }

    /**
     * Parse YAML string into a Map using SnakeYAML.
     */
    @Suppress("UNCHECKED_CAST")
    internal fun parseYaml(yamlContent: String): Map<String, Any> {
        return try {
            val result = yaml.load<Any>(yamlContent)
            (result as? Map<String, Any>) ?: emptyMap()
        } catch (e: Exception) {
            Timber.w(e, "SkillMdParser: YAML parse error, falling back to simple parser")
            simpleYamlParse(yamlContent)
        }
    }

    /**
     * Simple fallback YAML parser for basic key-value pairs.
     */
    private fun simpleYamlParse(content: String): Map<String, Any> {
        val result = mutableMapOf<String, Any>()
        for (line in content.lines()) {
            val trimmed = line.trim()
            if (trimmed.isBlank() || trimmed.startsWith("#")) continue

            val colonIndex = trimmed.indexOf(':')
            if (colonIndex > 0) {
                val key = trimmed.substring(0, colonIndex).trim()
                val value = trimmed.substring(colonIndex + 1).trim()

                if (value.startsWith("[") && value.endsWith("]")) {
                    // Inline array
                    result[key] = value.substring(1, value.length - 1)
                        .split(",")
                        .map { it.trim().removeSurrounding("\"").removeSurrounding("'") }
                        .filter { it.isNotBlank() }
                } else if (value == "true") {
                    result[key] = true
                } else if (value == "false") {
                    result[key] = false
                } else if (value.toIntOrNull() != null) {
                    result[key] = value.toInt()
                } else if (value.toDoubleOrNull() != null) {
                    result[key] = value.toDouble()
                } else {
                    result[key] = value.removeSurrounding("\"").removeSurrounding("'")
                }
            }
        }
        return result
    }

    /**
     * Build SkillMetadata from parsed YAML data.
     * Supports both kebab-case and camelCase field names (matching desktop).
     */
    @Suppress("UNCHECKED_CAST")
    internal fun buildMetadata(data: Map<String, Any>): SkillMetadata {
        val name = data["name"]?.toString() ?: "unknown"
        val displayName = (data["display-name"] ?: data["displayName"])?.toString() ?: name
        val description = data["description"]?.toString() ?: ""
        val version = data["version"]?.toString() ?: "1.0.0"
        val category = SkillCategory.fromString(
            data["category"]?.toString() ?: "general"
        )
        val author = data["author"]?.toString() ?: ""
        val tags = toStringList(data["tags"])
        val fileTypes = toStringList(
            data["supported-file-types"] ?: data["supportedFileTypes"] ?: data["file-types"] ?: data["fileTypes"]
        )
        val capabilities = toStringList(data["capabilities"])
        val userInvocable = (data["user-invocable"] ?: data["userInvocable"]) as? Boolean ?: true
        val hidden = data["hidden"] as? Boolean ?: false
        val handler = data["handler"]?.toString()
        val os = toStringList(data["os"]).ifEmpty { listOf("android", "win32", "darwin", "linux") }
        val executionMode = SkillExecutionMode.fromString(
            (data["execution-mode"] ?: data["executionMode"])?.toString() ?: "local"
        )
        val remoteSkillName = (data["remote-skill-name"] ?: data["remoteSkillName"])?.toString()

        // Agent Skills Open Standard fields
        val dependencies = toStringList(data["dependencies"])
        val cost = data["cost"]?.toString()
        val license = data["license"]?.toString() ?: ""
        val homepage = data["homepage"]?.toString() ?: ""
        val repository = data["repository"]?.toString() ?: ""

        // Input/output schema
        val inputSchema = parseParameters(data["input-schema"] ?: data["inputSchema"])
        val outputSchema = parseParameters(data["output-schema"] ?: data["outputSchema"])

        // Model hints
        val modelHints = (data["model-hints"] ?: data["modelHints"]) as? Map<String, Any> ?: emptyMap()

        // Gate conditions
        val gate = parseGate(data["gate"] ?: data["requires"])

        return SkillMetadata(
            name = name,
            displayName = displayName,
            version = version,
            description = description,
            category = category,
            author = author,
            tags = tags,
            fileTypes = fileTypes,
            capabilities = capabilities,
            userInvocable = userInvocable,
            hidden = hidden,
            inputSchema = inputSchema,
            outputSchema = outputSchema,
            modelHints = modelHints,
            dependencies = dependencies,
            cost = cost,
            license = license,
            homepage = homepage,
            repository = repository,
            gate = gate,
            handler = handler,
            os = os,
            executionMode = executionMode,
            remoteSkillName = remoteSkillName
        )
    }

    @Suppress("UNCHECKED_CAST")
    private fun toStringList(value: Any?): List<String> {
        return when (value) {
            is List<*> -> value.filterNotNull().map { it.toString() }
            is String -> if (value.isBlank()) emptyList()
            else value.split(",").map { it.trim() }.filter { it.isNotBlank() }
            else -> emptyList()
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseParameters(value: Any?): List<SkillParameter> {
        if (value == null) return emptyList()
        val list = value as? List<*> ?: return emptyList()

        return list.filterNotNull().mapNotNull { item ->
            val map = item as? Map<String, Any> ?: return@mapNotNull null
            SkillParameter(
                name = map["name"]?.toString() ?: return@mapNotNull null,
                type = map["type"]?.toString() ?: "string",
                description = map["description"]?.toString() ?: "",
                required = map["required"] as? Boolean ?: false,
                default = map["default"]
            )
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun parseGate(value: Any?): SkillGate? {
        val map = value as? Map<String, Any> ?: return null
        return SkillGate(
            platform = toStringList(map["platform"]).ifEmpty { null },
            minSdk = (map["minSdk"] ?: map["min-sdk"]) as? Int,
            requiredPermissions = toStringList(map["permissions"] ?: map["requiredPermissions"]).ifEmpty { null },
            requiredBinaries = toStringList(map["bins"] ?: map["requiredBinaries"]).ifEmpty { null },
            requiredEnv = toStringList(map["env"] ?: map["requiredEnv"]).ifEmpty { null }
        )
    }

    /**
     * Extract code examples from Markdown body (fenced code blocks after "## Examples").
     */
    private fun extractExamples(body: String): List<String> {
        val examples = mutableListOf<String>()
        val lines = body.lines()
        var inExamplesSection = false
        var inCodeBlock = false
        val currentBlock = StringBuilder()

        for (line in lines) {
            val trimmed = line.trim()

            if (trimmed.startsWith("## Example")) {
                inExamplesSection = true
                continue
            }

            // Exit examples section on next h2
            if (inExamplesSection && trimmed.startsWith("## ") && !trimmed.startsWith("## Example")) {
                inExamplesSection = false
            }

            if (inExamplesSection) {
                if (trimmed.startsWith("```")) {
                    if (inCodeBlock) {
                        // End of code block
                        examples.add(currentBlock.toString().trim())
                        currentBlock.clear()
                        inCodeBlock = false
                    } else {
                        inCodeBlock = true
                    }
                } else if (inCodeBlock) {
                    currentBlock.appendLine(line)
                }
            }
        }

        return examples
    }
}
